Kwxm/saturated builtins (SCP-250, SCP-533).

kwxm opened the pull request:
This updates `language-plutus-core` by making application of built-in functions saturated.  This is just a draft PR so that I can get comments to make sure that it's going in roughly the right direction. It's still pretty messy in places, so I don't think a full review is necessary just yet.

More or less everything in `language-plutus-core` appears to work (apart from the golden tests; also prettyprinting does something, but it's definitely not perfect).  The two major things I had difficulty with are dynamic builtin application and typechecking. I haven't made any attempt to adapt `plutus-ir` or `plutus-tx` yet: I think that should wait until `language-plutus-core` is OK.

### Dynamic application
Dynamic builtin evaluation is broken: in `MkPlc`, `dynamicBuiltinNameAsTerm` no longer makes sense and just raises an error.  I've managed to remove most uses of this, but it's used in `OnChain.hs` and I think it may require a fair amount of reworking to fix that.  It'd probably be best for @effectfully to take a look at that.

### Typechecking
There are also some difficulties with typechecking, which might require us to change the representation of builtin applications.

The basic change in this PR was to modify the `Term` type by replacing `Builtin` with
```
    ApplyBuiltin ann BuiltinName [Type tyname uni ann] [Term tyname name uni ann]
```
There's a problem with this is that there are a number of places where we need to know that the lengths of the lists are correct, and _you can't tell this from the type_.  Types for builtins are represented in two ways: using the `Type` type and using the `TypeScheme` type.  The problem is that we have a mapping from TypeSchemes to Types, but it's not injective.  If you have a builtin of Type `int → int -> int` then you don't know whether it takes two arguments and returns an integer, or one argument and returns a function (but these _do_ have different TypeSchemes.  This makes it hard to work out if a builtin's been applied to the right number of arguments.  I've added a temporary workaround for that in `TypeCheck.Internal`, but it's very messy.  Things are better when the TypeScheme is available, but there's still some work required to take these apart (see `Denotation.hs`). Maybe we could make TypeSchemes available everywhere (probably requiring more monads) and provide some machinery for decomposing them? [Incidentally, is the `TypeScheme` type more general than is required?  It allows us to represent things with types `A →B →C` where there are `forall`s in `B` or `C`, and it's not clear to me if we really need this].

Things could probably be made better if builtins carried their arities around with them. I think there are at least two ways to do this: (1) just add two integer fields to the `BuiltinName` type; and (2) use something like `Vec` to  do it in the type system.  I've never attempted to use `Vec` so it's not clear to me how difficult that would be (and parsing and deserialisation might be tricky).  In the first instance (1) might be better.  However, maybe there's some better way to deal with this that I haven't thought of.  Also, instead of trying to check the arities in the typechecker it might be better to do it in a preliminary sanity check.  

As mentioned above, this is still very messy. There are quite a number of comments beginning `FIXME` flagging up things that I need to come back to.

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -303,17 +305,105 @@ unfoldFixOf pat arg k = do
             ]
 
 -- | Infer the type of a 'Builtin'.
+-- FIXME: Annoyingly we have to supply an annotation here for the error if the name isn't found
```
I'd say it's fine, we do that in many places.

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -303,17 +305,105 @@ unfoldFixOf pat arg k = do
+-- eg  o : forall a b c: (b->c) -> (a->b) -> (a->c).
+-- Things would be a lot easier if (a) we had the type schema for the
+-- builtin at this point or (b) builtin names carried their arities
+-- about with them.
```
You have the type scheme of a builtin at any point. For a dynamic builtin it amounts to performing a lookup in the environment. And for a static builtin you can call [`withTypedBuiltinName`](https://github.com/input-output-hk/plutus/blob/25eea28df681f74fd0ce4e8e483b619e52a38220/language-plutus-core/src/Language/PlutusCore/Constant/Name.hs#L60). That should change the way you do type checking.

effectfully commented on language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/List.hs:
```diff
@@ -251,19 +250,33 @@ enumFromTo = runQuote $ do
-        add = builtin () (BuiltinName () AddInteger)
-    return
-        . mkIterApp () (mkIterInst () foldList [int, int])
-        $ [ add , mkConstant @Integer () 0]
```
Why change all of that? Just use eta-expansion to turn a builtin into a `Term` and avoid all the trouble.

kwxm replied:
I think I did that before implementing the eta-expansion code.  I'll probably just leave it since it works.

effectfully commented:
> Dynamic builtin evaluation is broken: in `MkPlc`, `dynamicBuiltinNameAsTerm` no longer makes sense and just raises an error. I've managed to remove most uses of this, but it's used in `OnChain.hs` and I think it may require a fair amount of reworking to fix that.

Feel free to remove `OnChain.hs`, I don't think it's needed for anything. I've been meaning to do that anyway.

kwxm commented on language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs:
```diff
@@ -330,46 +350,55 @@ withScopedArgIn funVarEnv argVarEnv arg            k = do
+    :: [Closure uni]           -- original arguments
+    -> [WithMemory Value uni]  -- adjusted arguments
+    -> VarEnv uni              -- enviroment containing bindings for new variables
+    -> CekM uni ([WithMemory Value uni], VarEnv uni)
```
OK, I'll look at that.  I went for a direct implementation initially because CPS isn't good for my brain.

effectfully replied:
That's just my personal preference, I don't think it's too important.

effectfully commented on language-plutus-core/generators/Language/PlutusCore/Generators/Interesting.hs:
```diff
@@ -51,21 +52,17 @@ type TermGen uni a = Gen (TermOf uni a)
 -- > ifThenElse {integer -> integer -> integer} (lessThanInteger i j) addInteger subtractInteger i j
 -- >     == if i < j then i + j else i - j
-genOverapplication :: Generatable uni => TermGen uni Integer
+genOverapplication :: Generatable uni => TermGen uni Integer  -- FIXME: Doesn't make sense any more???
```
Why not? The stdlib is mostly there for tests, for example (apart from the `Type.hs` module, n-ary tuples and maybe a few more things).

kwxm replied:
> I think @michaelpj also eta-expands builtins in Plutus Tx. Is that right?

Yes, but we're hoping that will go away: see https://jira.iohk.io/browse/SCP-480.

When I said "we don't want to use it in general" I was worried about the possibility of ending up with inefficient wrapped builtin applications in places where you could just have a direct application, and we should be trying to generate those wherever possible (and up front, rather than as a later optimisation).  What happens in the PlutusTx output is a different matter: I think it's OK have a later optimisation phase there.

kwxm commented:
> Feel free to remove `OnChain.hs`, I don't think it's needed for anything. I've been meaning to do that anyway.

Aha!  That didn't occur to me at all: I thought it was still in use.  Getting rid of it should simplify things considerably.

effectfully commented on language-plutus-core/generators/Language/PlutusCore/Generators/Internal/Denotation.hs:
```diff
@@ -72,11 +78,78 @@ typeSchemeResult (TypeSchemeAllType _ schK) = typeSchemeResult $ schK Proxy
+                    let text = pack $ symbolVal @text Proxy
+                        uniq = fromIntegral $ natVal @uniq Proxy  -- Can we just make a new unique here?
+                        a    = TyName $ Name text $ Unique uniq
+                        tvd = TyVarDecl () a (Type ()) -- FIXME: will we ever need a higher kind here?
```
No, it's called `AllType` precisely because it can only bind a `*`.

effectfully commented on language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs:
```diff
@@ -242,34 +259,34 @@ computeCek con t@(Var _ varName)   = do
-        throwingWithCause _MachineError NonPrimitiveInstantiationMachineError $ Just (void fun)
+returnCek (FrameApplyBuiltin env bn tys closures terms : ctx) value = do
+    varEnv <- getVarEnv
+    let closures' = Closure varEnv value : closures -- Save the closure for the argument we've just computed.
```
If we go with CPS, then we can store a continuation instead of a list of closures, because that is how the list of closures is going to be used down the control flow anyway.

kwxm replied:
I don't think we even need CPS.  I think we can do it all with the frames (this is what the comment at the start of the definition of `getScopedArgs` was alluding to).  The frame for builtin application has to save the environment, `env0`, say that the application's being evaluated in because it needs it for each of that arguments.  When you compute an argument down to a value the `return` phase creates a closure which gets saved in a new frame and then it goes on to deal with the next argument.  I think that when you're returning a newly-computed argument value you can look at it to see if it's a constant or not, and if it's not constant, bind it to a variable and then extend `env0` with the new binding and use that to compute the next argument.  Now maybe that's not quite right, since the next argument is being computed in a larger environment than it should be.  I don't think that matters, because we've used new variables, but even if it does matter we can still accumulate the extended environment in the frame and at the end use that to carry on with the computation.  I didn't want to do that right away because while I was implementing this it was less complicated to deal with the arguments separately (and test that that method worked), but I think it should work.

effectfully commented on language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs:
```diff
@@ -330,46 +350,55 @@ withScopedArgIn funVarEnv argVarEnv arg            k = do
+    -- If there are too many/few types this should be caught by the typechecker.
+    let term = ApplyBuiltin () bn (fmap void tys) (fmap (void . _closureValue) args) -- For error messages
+    in do
+      -- Do we have to be more careful for dynamic names?
```
Do you have something in mind?

kwxm replied:
I'm not sure: I think that was just to remind me to go back and think more carefully.

effectfully commented on language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs:
```diff
@@ -242,34 +259,34 @@ computeCek con t@(Var _ varName)   = do
+    let closures' = Closure varEnv value : closures -- Save the closure for the argument we've just computed.
+    case terms of
+      []       -> applyBuiltin env () ctx bn tys (reverse closures') -- We've accumulated the argument closures in reverse
+      arg:args -> withVarEnv env $ computeCek (FrameApplyBuiltin env bn tys closures' args : ctx) arg
```
Saving the environment of a builtin to restore it each time we evaluate a new argument makes sense to me.

kwxm commented on language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs:
```diff
@@ -242,34 +259,34 @@ computeCek con t@(Var _ varName)   = do
+    varEnv <- getVarEnv
+    let closures' = Closure varEnv value : closures -- Save the closure for the argument we've just computed.
+    case terms of
+      []       -> applyBuiltin env () ctx bn tys (reverse closures') -- We've accumulated the argument closures in reverse
```
Yes, but that's a bit fragile because quite a lot goes on between the two reversals, and the fact that everything ends up in the right order is a lucky consequence of the fact that you call two things that reverse their arguments.  I've got into the habit of reversing accumulators at the first possible opportunity because if you forget to do it you can end up with confusing bugs.

Anyway, this will go away if we just merge the functionality of `getScopedArgs` into the frames.

kwxm commented on language-plutus-core/generators/Language/PlutusCore/Generators/AST.hs:
```diff
@@ -112,7 +112,9 @@ genTerm = simpleRecursive nonRecursive recursive where
     recursive = [absGen, instGen, lamGen, applyGen, unwrapGen, wrapGen]
-    nonRecursive = [varGen, Constant () <$> genConstant, Builtin () <$> genBuiltin, errorGen]
+    nonRecursive = [varGen, Constant () <$> genConstant, errorGen] -- FIXME, Builtin () <$> genBuiltin]
+-- Tricky.  How do we get the instantion/arity right for builtins?
```
I was worried that if we have, say, a builtin with two type arguments and three term arguments and just generate random lists of arguments then the probability of getting a sensible example might be quite low.  I will put that generator back in though.

kwxm replied:
Seems OK.

jmchapman commented:
I am surprised by the ambiguity between `int → int → int` and `int → (int → int)`. I was expecting saturation to mean the return type should be a ground type. So, in the this case two arguments should be given. I guess this means that `→` shouldn't be allowed in a return type of a builtin and also `forall` shouldn't be allowed either. Having `forall` would also be weird as it would allow you to sometimes have have type args appearing after term args and sometimes not.

effectfully commented on language-plutus-core/test/Evaluation/Machines/Budget/Examples/InterList/FoldrInterList.plc.golden:
```diff
@@ -1,8 +1,14 @@
-( (Right (/\(a0 :: *) -> /\(b0 :: *) -> /\(r :: *) -> \(f : a0 -> b0 -> r -> r) -> \(z : r) -> (/\(a :: *) -> /\(b :: *) -> \(f : (a -> b) -> a -> b) -> (/\(a :: *) -> \(s : (\(a :: *) -> ifix (\(self :: * -> *) -> \(a :: *) -> self a -> a) a) a) -> unwrap s s) {a -> b} (iwrap (\(self :: * -> *) -> \(a :: *) -> self a -> a) (a -> b) (\(s : (\(a :: *) -> ifix (\(self :: * -> *) -> \(a :: *) -> sel ...
+( (Left An error has occurred:  User error:
+The limit ({ cpu: 1000
+| mem: 1000
+}) was reached by the execution environment. Final state: ({ cpu: 0
+| mem: 1090
+})
+Caused by: (/\(a0 :: *) -> /\(b0 :: *) -> /\(r :: *) -> \(f : a0 -> b0 -> r -> r) -> \(z : r) -> (/\(a :: *) -> /\(b :: *) -> \(f : (a -> b) -> a -> b) -> (/\(a :: *) -> \(s : (\(a :: *) -> ifix (\(self :: * -> *) -> \(a :: *) -> self a -> a) a) a) -> unwrap s s) {a -> b} (iwrap (\(self :: * -> *) -> \(a :: *) -> self a -> a) (a -> b) (\(s : (\(a :: *) -> ifix (\(self :: * -> *) -> \(a :: *) -> s ...
 , ({ tally: ({ BAST causes ({ cpu: 0
-| mem: 345
+| mem: 1090
```
@kwxm @reactormonk I guess you'll need to talk about costing for saturated builtins?

kwxm replied:
> Still going, haven't got to the typechecker yet, but here's what I've got so far.

I wasn't expecting anyone to review it just yet (still tidying up a few things), but go ahead anyway.

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -328,11 +438,17 @@ inferTypeM (Constant _ (Some (ValueOf uni _))) =
+-- G !- ({builtin bn A_1, ..., A_m} : [A_1,...,A_m/a_1,...,a_m]C)
+inferTypeM (ApplyBuiltin ann bn tyargs args) = do
+  TypeComponents typeVars argTypes rtype <- inferTypeComponentsOfBuiltinM ann bn
+  rtype' <- checkBuiltinAppl ann bn typeVars tyargs argTypes args rtype
```
I'd inline the definition of `checkBuiltinAppl` as it's just

1. check types
2. check terms
3. return the final type

which all seem to be very different things. Pulling the whole `do` into a separate function would also be sensible, then inlining `checkBuiltinAppl` would look even better. Right now `checkBuiltinAppl` receives way too many arguments loosely related to each other.

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -302,18 +326,104 @@ unfoldFixOf pat arg k = do
+    -> TypeCheckM uni ann (Normalized (Type TyName uni ()))
+checkBuiltinAppl ann bn typeVars typeArgs argTypes args rtype = do
+  mapping <- checkBuiltinTypeArgs ann bn typeVars typeArgs
+  checkBuiltinTermArgs mapping ann bn argTypes args
```
We could remove the `mapping` argument by discharging the `mapping` into `argTypes` via `map (substMapping mapping)`. Then `checkBuiltinTermArgs` doesn't need to know anything about mappings and can just directly type check arguments, which makes the code less tightly coupled.

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -302,18 +326,104 @@ unfoldFixOf pat arg k = do
+    check mapping ann bn (tyname:tynames) (tyarg:tyargs) = do
+       checkKindM  (typeAnn tyarg) tyarg (Type ()) -- NB: All of the type variables in TypeSchemes have kind *
+       tyarg' <- normalizeType (void tyarg)
+       check ((tyname,tyarg'):mapping) ann bn tynames tyargs
```
We'll never need to do anything with `mapping` apart from feeding it to `substMapping`, right? In that case I'd construct the substituter directly just not to introduce an indirection. This way it would be also clearer whether we substitute types left to right or right to left (I don't think it's important from the correctness point of view, but it would make it easier to understand the whole thing).

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -139,16 +142,37 @@ withTyVar name = local . over tceTyVarKinds . insertByName name
--- | Look up a 'DynamicBuiltinName' in the 'DynBuiltinNameTypes' environment.
+-- | Look up a 'DynamicBuiltinName' in the 'DynBuiltinNameTypes'
+-- environment and return its type components.
 lookupDynamicBuiltinNameM
```
Do we still need this function?

kwxm replied:
Apparently not.

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -302,18 +326,104 @@ unfoldFixOf pat arg k = do
+          -- ^ annoying: we have to normalise during subsitution, but
+          -- then we need a non-normalised type to perform subsequent
+          -- subsitutions on.
+    ty' <- foldM f (void ty) mapping
```
I think a better way would be to normalize `ty` passed to `foldM` and then instead of unwrapping the result you'll need to unwrap the argument in `f`, which is better because the result is going to be `Normalized`, so no additional `normalizeTypeM ty'` will be required.

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -302,18 +326,104 @@ unfoldFixOf pat arg k = do
+checkBuiltinTypeArgs = check [] where
+    check mapping ann bn (tyname:tynames) (tyarg:tyargs) = do
+       checkKindM  (typeAnn tyarg) tyarg (Type ()) -- NB: All of the type variables in TypeSchemes have kind *
+       tyarg' <- normalizeType (void tyarg)
```
We need a note explaining how global uniqueness is preserved. Did you take any special care for handling global uniqueness? It looks correct to me, because you never wrap anything with `Normalized` explicitly and functions like `normalizeType` and `substNormalizeTypeM` ensure that global uniqueness is preserved.

effectfully commented on language-plutus-core/src/Language/PlutusCore/Core/Instance/Eq.hs:
```diff
@@ -117,18 +113,22 @@ eqTermM (Var _ name1) (Var _ name2) =
+    guard $ length tys1 == length tys2
+    zipWithM_ eqTypeM tys1 tys2
+    guard $ length args1 == length args2
+    zipWithM_ eqTermM args1 args2
```
Actually, I have a feeling that having

```haskell
data OneIsLonger = LeftIsLonger | RightIsLonger

zipExact :: [a] -> [b] -> Either OneIsLonger [(a, b)]
```

we can make this code and the type checker's code nicely readable. Especially the type checke, because there we won't need some explicit recursion then.

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -302,18 +326,104 @@ unfoldFixOf pat arg k = do
+    -> TypeCheckM uni ann (Normalized (Type TyName uni ()))
+substMapping mapping ty = do
+    let f ty1 (name, newty) = do
+          ty2 <- substNormalizeTypeM newty name ty1
```
`substNormalizeTypeM` is defined as

```
substNormalizeTypeM ty name = withExtendedTypeVarEnv name ty . normalizeTypeM
```

and so calling it iteratively is not efficient. Instead, we should provide a primitive for binding a map of, well, mappings in `Normalize.Internal`, expose it and use it here.

effectfully commented on language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs:
```diff
@@ -332,75 +363,87 @@ and transition it into
+                 -> [Value TyName Name uni ExMemory]
+                 -> CekM uni (Term TyName Name uni ())
+applyBuiltin finalEnv ctx bn tys args =
+    -- If there are too many/few types this should be caught by the typechecker.
```
Rather, by the builtin application machinery (from `Language.PlutusCore.Constant.Apply`), because we won't run the type checker on the chain.

effectfully commented on language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs:
```diff
@@ -332,75 +363,87 @@ and transition it into
+   Currently each argument is evaluated in initialEnv.  However, we might
+   be able to get away with using finalEnv instead, since every variable
+   in finalEnv\initialEnv is fresh and thus should not appear in any of the
+   input terms.
```
I like the idea of having a separate `finalEnv` as it better mirrors how things get computed.

Great note, btw.

effectfully commented on language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs:
```diff
@@ -273,34 +292,45 @@ computeCek con t@(Var _ varName)   = do
-        throwingWithCause _MachineError NonPrimitiveInstantiationMachineError $ Just (void fun)
+returnCek (FrameApplyBuiltin initialEnv finalEnv bn tys values terms : ctx) value = do
+    -- See Note [Built-in application] for an explanation of what's going on here.
+    case value of
+      Constant{} -> do
+          let values' = value:values
+          case terms of
+            []       -> applyBuiltin finalEnv ctx bn tys (reverse values') -- We've accumulated the argument closures in reverse
+            arg:args -> withVarEnv initialEnv $ computeCek (FrameApplyBuiltin initialEnv finalEnv bn tys values' args : ctx) arg
```
I'd perhaps return `values'` and `newFinalEnv` (why not `finalEnv'` or `newValues` then?) from `case value of` and put a single `case terms of` below `case value of`. Otherwise we duplicate some non-trivial code, which is not very good.

kwxm replied:
Ah yes, I'd meant to go back and look at that. It's a lot neater now.

effectfully commented on language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs:
```diff
@@ -239,29 +246,41 @@ substitution, anything).
+    -- FIXME (FAO @reactormonk): I've just re-used BApply for costing
+    -- the use of FrameApplyBuiltin.  Maybe we should add a new
+    -- constructor to `ExBudgetCategory` for that.
+    case args of
+        [] -> do
+            spendBudget BApply t (ExBudget 1 1)
+            varEnv <- getVarEnv
+            applyBuiltin varEnv ctx bn tys []
```
I think we don't really support this case? At least costing doesn't work correctly here IIRC. I think we could throw an error here and think about this case later.

kwxm replied:
> A thought: we could use `NonEmpty` for the builtin arguments and give an error way back in the parser if there's no arguments at all.

OK,  I'll try that.  I did wonder if we could encode the type and term arities at the type level so that we could _never_ have an application that was incorrectly instantiated/applied, but I thought that was too ambitious at this point.  I'm also not sure how that would interact with CBOR decoding, for example.

michaelpj replied:
Yeah, matching arity sounds annoying, but non-emptiness is a pretty easy thing to add if you know it holds.

Decoding is a good example, though: if we made nullary builtins work, it would be nice if we didn't have to change the serialization format...

kwxm replied:
I spent quite a long time on this and then gave up.  It requires changes to a lot of code, including the prettyprinter, the typechecker and so on. It was just making a horrible mess of everything, so I've made the parser and the evaluators (but not the CBOR decoder, yet) reject nullary builtin applications.  I might revisit this later, but not right now.

kwxm replied:
I made an issue about nullary builtins at SCP-779.

effectfully commented on language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs:
```diff
@@ -273,34 +292,45 @@ computeCek con t@(Var _ varName)   = do
-returnCek (FrameTyInstArg ty : con) fun = instantiateEvaluate con ty fun
-returnCek (FrameApplyArg argVarEnv arg : con) fun = do
+returnCek (FrameTyInstArg _ : ctx) fun =
+    case fun of
```
(I don't mind inlining the match)

kwxm replied:
I think I originally had separate matches but then changed it because there are quite a lot of cases in `returnCek` and I thought it'd make it easier to see what's going on if there was exactly one top-level match for each kind of frame.

effectfully commented on language-plutus-core/src/Language/PlutusCore/Eq.hs:
```diff
@@ -179,3 +179,4 @@ eqNameM name1 name2 = do
 -- | Check equality of things having an 'Eq' instance.
 eqM :: Eq a => a -> a -> EqRename ren
 eqM x1 x2 = guard $ x1 == x2
```
Indeed! We don't even need to do zipping then, just call `==` over the lists.

effectfully replied:
(here was a stupid comment left by me)

effectfully commented on language-plutus-core/src/Language/PlutusCore/Constant/Embed.hs:
```diff
@@ -0,0 +1,59 @@
+      Nothing -> Prelude.error $ "Malformed type scheme for " ++ show name ++ " in embedBuiltinNameInTerm"
+      -- FIXME: we should really have a proper error here, but that involves running a *lot* of stuff in a monad
+      Just (TypeComponents tynames types _) ->
+          runQuote $ do
```
The resulting term is supposed to be closed, right? In that case we don't need `MonadQuote` (e.g. we call `runQuote` everywhere in stdlib, examples etc). Global uniqueness does not get preserved this way, but it's just like in `typeSchemeToType` -- we ensure global uniqueness in the type checker by calls to `normalizeType` (or `liftDupable` sometimes).

kwxm commented on language-plutus-core/generators/Language/PlutusCore/Generators/Interesting.hs:
```diff
@@ -47,25 +47,21 @@ import qualified Hedgehog.Range                           as Range
+-- | Generates application of a builtin
 --
--- > ifThenElse {integer -> integer -> integer} (lessThanInteger i j) addInteger subtractInteger i j
+-- > ifThenElse {integer} (lessThanInteger i j) (addInteger i j) (subtractInteger i j)
```
Done.

kwxm commented on language-plutus-core/src/Language/PlutusCore/Constant/Embed.hs:
```diff
@@ -0,0 +1,59 @@
+          -- This makes the unique explicit in the variable name.  If
+          -- you don't do this then `plc example` outputs examples
+          -- where builtins are applied to multiple arguments with the
+          -- same name.
```
:+1:

kwxm commented on language-plutus-core/src/Language/PlutusCore/Constant/Function.hs:
```diff
@@ -51,3 +53,60 @@ insertDynamicBuiltinNameDefinition
+   TypeScheme to a TypeComponents object.
+-}
+data TypeComponents uni = TypeComponents { tcTypeVars   :: [TyName]
+                                         , tcArgTypes   :: [Type TyName uni ()]
```
I think I tried that as well, but it was a bit annoying either way (if you make a VarDecl, you have to take it apart sometimes to get the types out).  I think @effectfully suggested that he might re-work the TypeScheme stuff at some time in the future, so with luck all this stuff might go away.

effectfully replied:
> Although then you have to come up with a name which is probably a pain.

We could use `arg_1`, `arg_2` etc, but generating names is scary, so it's nice to avoid that when possible.

>  I think @effectfully suggested that he might re-work the TypeScheme stuff at some time in the future, so with luck all this stuff might go away.

I'm not sure if there's a better way of doing type checking even with reworked `TypeScheme`.

kwxm commented on language-plutus-core/src/Language/PlutusCore/Constant/Embed.hs:
```diff
@@ -0,0 +1,59 @@
+                tyVarDecls = map mkTyVarDecl tynames
+            pure $ mkIterTyAbs tyVarDecls (mkIterLamAbs varDecls (ApplyBuiltin () name tyArgs termArgs))
+
+embedStaticBuiltinNameInTerm :: TypeScheme uni args res -> StaticBuiltinName -> Term TyName Name uni ()
```
OK, I did that.  Using `withTypedBuiltinName` meant that I had to add ` (GShow uni, GEq uni, DefaultUni <: uni)` constraints to `embedStaticBuiltinNameInTerm` and a number of things in `Denotation.hs` though. In the end it didn't matter because those constraints always held where the functions were ultimately being called.

kwxm commented on language-plutus-core/plutus-ir/Language/PlutusIR/Parser.hs:
```diff
@@ -120,7 +121,6 @@ inBraces = between lbrace rbrace
 
 reservedWords :: [T.Text]
 reservedWords =
-    map display PLC.allBuiltinNames ++
```
Why remove the built-in names from the list of reserved words?  Because the compiler wraps builtins in lambdas with the same names and then if you print this out you can't re-parse it if the names are keywords. I removed keywords from the main parser completely a while ago but didn't want to start messing about with the PIR parser at that point. Once I've got this PR out of the way I'll have a look at the thing about putting type names into literal constants and when I'm doing that I may update the PIR parser as well.

effectfully replied:
I see, thanks for the explanation.

kwxm commented on language-plutus-core/src/Language/PlutusCore/CBOR.hs:
```diff
@@ -161,24 +159,24 @@ instance Serialise ann => Serialise (Version ann) where
-        a (KindArrowF ann k k') = fold [ encodeConstructorTag 1, encode ann, k , k' ]
+    encode = \case
+        Type ann           -> encodeConstructorTag 0 <> encode ann
+        KindArrow ann k k' -> encodeConstructorTag 1 <> encode ann <> encode k <> encode k'
```
I was forced to remove the recursion schemes from the CBOR code because they caused a somewhat subtle bug when you're encoding/decoding lists.  Recall that cborg introduces an intermediate monoidal type `Encoding` and chunks of data are converted to this and then all concatenated into a single thing when you finally want to output the encoded version.  When it's outputting a list, it writes a tag saying "here's the start of a list", then writes the encoded entries one after the other, then writes a tag saying "end of list".  When you use a catamorphism, it maps the encoding function over the list, encoding all the entries, and then **it concatenates all of the `Encoding` objects into a single object, omitting the markers at the beginning and end**.  If you write this to disk and then try to read it back in, it'll fail. This is a case where you really don't want to be functorial!

kwxm commented on language-plutus-core/src/Language/PlutusCore/Lexer/Type.hs:
```diff
@@ -141,3 +144,32 @@ newIdentifier str = do
+
+
+toStaticBuiltinName :: T.Text -> Maybe StaticBuiltinName
+toStaticBuiltinName = \case
```
@effectfully But somewhere we have to convert the syntactic names of the builtins into the relevant constructors (which are what are stored in `allBuiltinNames`).  I had a look, but I can't see that happening anywhere else.  I don't quite see what you mean here.

effectfully replied:
You can do

```haskell
import qualified Data.Map.Strict as Map

allBuiltinNamesMap :: Map.Map String BuiltinName
allBuiltinNamesMap = Map.fromList $ map (\x -> (show $ pretty x, x)) allBuiltinNames
```

and look up syntactic names in `allBuiltinNamesMap`, right?

effectfully replied:
You can do

```haskell
import qualified Data.Map.Strict as Map

allBuiltinNamesMap :: Map.Map String BuiltinName
allBuiltinNamesMap = Map.fromList $ map (\x -> (show $ pretty x, x)) allBuiltinNames
```

and look up syntactic names in `allBuiltinNamesMap`, right?

kwxm replied:
That did cross my mind, but it almost seems the wrong way round:  it'd be quite hard to work out what the concrete syntax for the builtins is without digging quite deeply into the code.  

Having said that, I suppose we'll _have_ to do something like this for dynamic builtins, which will be important if we get to a point where that's all we have.  The parser should somehow be provided with a list of the builtins that are going to be available at execution time so that it can complain if the code contains something unexpected, so that'll have to be derived from a `DynamicBuiltinMeanings` structure or something similar.

effectfully replied:
> That did cross my mind, but it almost seems the wrong way round: it'd be quite hard to work out what the concrete syntax for the builtins is without digging quite deeply into the code.

The concrete syntax for builtins is the one that they get printed as. That's the point of not introducing two different syntaxes which happen to match.

> Having said that, I suppose we'll have to do something like this for dynamic builtins, which will be important if we get to a point where that's all we have. The parser should somehow be provided with a list of the builtins that are going to be available at execution time so that it can complain if the code contains something unexpected, so that'll have to be derived from a DynamicBuiltinMeanings structure or something similar.

Yep.

kwxm replied:
> The concrete syntax for builtins is the one that they get printed as.
I'm not entirely convinced by that: I could equally well say that the concrete syntax for builtins should be the one they get read as!  I think what disturbs me a bit here is that the concrete name for a builtin is rather loosely coupled to the builtin itself via the `Pretty` typeclass.  If you're looking at the code trying to see what the textual form of a builtin is then you have to locate the `Pretty` instances, which are buried away in a file that's far from easy to find, and if you're trying to do it programatically you have to do it indirectly, by calling `show . pretty`  I think I'd be happier if the definitive string representation of a builtin was included directly in the object defining the builtin itself (maybe the `DynamicBuiltinNameDefinition` for example), where it could be accessed easily by both the prettyprinter and the parser (and also by people looking at the code).  Maybe we could think about doing that if and when we get rid of static builtins.

Anyway, never mind: this is perhaps a bit of a theological argument.  I'll get rid of that table as you suggested.

effectfully replied:
> I'm not entirely convinced by that: I could equally well say that the concrete syntax for builtins should be the one they get read as!

That would also be fine. Anything is fine as long as there's a single source of truth.

>  If you're looking at the code trying to see what the textual form of a builtin is then you have to locate the `Pretty` instances, which are buried away in a file that's far from easy to find, and if you're trying to do it programatically you have to do it indirectly, by calling `show . pretty`

I think it's fairly straightforward to get what's happening by looking at the definition of `toStaticBuiltinName` defined in terms of `allBuiltinNamesMap`.

> Maybe we could think about doing that if and when we get rid of static builtins.

If we keep dynamic builtins as mere `Text`, then we won't need to go through `Pretty`. If we introduce an additional type variable, then going through `Pretty` still seems sensible and convenient to me.

kwxm commented on language-plutus-core/src/Language/PlutusCore/Subst.hs:
```diff
@@ -154,7 +154,7 @@ fvTerm = cata f
     f (IWrapF _ _ _ t)  = t
     f ConstantF{}       = Set.empty
-    f BuiltinF{}        = Set.empty
+    f ApplyBuiltinF{}   = Set.empty
```
I have absolutely no idea why I wrote that! Thanks for pointing that out.

kwxm replied:
Is strictness really of any importance here? `Set.unions` does the job nicely, but it's defined in terms of `foldl`.

effectfully replied:
Seems like a bug in their docs. I see this in source code:

```haskell
unions :: (Foldable f, Ord a) => f (Set a) -> Set a
unions = Foldable.foldl' union empty
```

`Set.unions` is indeed better than what I suggested.

kwxm commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -328,11 +438,17 @@ inferTypeM (Constant _ (Some (ValueOf uni _))) =
+-- G !- M_i : D_i
+-- D_i == [A_1,...,A_m/a_1,...,a_m]B_i
+----------------------------------------------------------------
+-- G !- ({builtin bn A_1, ..., A_m} : [A_1,...,A_m/a_1,...,a_m]C)
```
I was planning to go back and sort that out, but you beat me to it!  The comments are helpful though.

kwxm commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -328,11 +438,17 @@ inferTypeM (Constant _ (Some (ValueOf uni _))) =
+-- G !- ({builtin bn A_1, ..., A_m} : [A_1,...,A_m/a_1,...,a_m]C)
+inferTypeM (ApplyBuiltin ann bn tyargs args) = do
+  TypeComponents typeVars argTypes rtype <- inferTypeComponentsOfBuiltinM ann bn
+  rtype' <- checkBuiltinAppl ann bn typeVars tyargs argTypes args rtype
+  pure rtype'
```
Yes. I think I did that just to make it clear what was actually being returned.

kwxm commented on language-plutus-core/test/Evaluation/ApplyBuiltinName.hs:
```diff
@@ -16,12 +16,14 @@ import           Language.PlutusCore.Pretty
 
+-- FIXME: add some tests for under/overapplication?  We used to have
+-- ConstAppStuck and the tests here checked for that with
+-- underapplication, but now we have exceptions instead.
```
Where would be a suitable place to put such tests then?  The structure of the tests is quite complicated and I'm not sure what should go where.  (If I do add some tests I might do it in a separate small PR, just to get this one out of the way.)

effectfully replied:
Not sure. Perhaps a new module in `test/Evaluation`?

kwxm commented on plutus-tx-plugin/src/Language/PlutusTx/Compiler/Builtins.hs:
```diff
@@ -397,63 +391,72 @@ errorTy = do
+{-
+  eqInteger :   lam x . lam y . (convertor (builtin eqInteger [] [x, y]))
+             -> lam x . lam y . (lam b plcBoolTy . (ifThenElse [haskellBoolTy] [b, True, False])) <applied to> (builtin eqInteger [] [x, y]) 
+-}
```
A note to myself which I hadn't got round to removing before you started reviewing.

kwxm commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -102,7 +104,8 @@ type VarTypes uni = UniqueMap TermUnique (Dupable (Normalized (Type TyName uni (
 data TypeCheckConfig uni = TypeCheckConfig
-    { _tccDynamicBuiltinNameTypes :: DynamicBuiltinNameTypes uni
+    { _tccDynamicBuiltinNameTypes    :: DynamicBuiltinNameTypes uni
+    , _tccDynamicBuiltinNameMeanings :: DynamicBuiltinNameMeanings uni
```
Oooh, you're right!  That's good: I was able to delete a surprising amount of code thanks to that.

kwxm commented on language-plutus-core/src/Language/PlutusCore/View.hs:
```diff
@@ -29,41 +25,13 @@ data IterApp head arg = IterApp
-
 -- | An iterated application of a 'Term' to a list of 'Term's.
 termAsTermIterApp :: Term tyname name uni a -> TermIterApp tyname name uni a
 termAsTermIterApp = go [] where
```
Nope. Removed.

kwxm commented:
I've lost track a bit, but I think I've addressed most of the comments so far.  This should be ready for another look.  Remember that we haven't decided whether or not we actually want this though!

michaelpj commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -302,18 +316,62 @@ unfoldFixOf pat arg k = do
+      Left LeftLonger  -> throwError $ BuiltinTypeErrorE ann (BuiltinUnderInstantiated bn)
+      Left RightLonger -> throwError $ BuiltinTypeErrorE ann (BuiltinOverInstantiated  bn)
+      Right namesAndTypes ->  do
+        mapM_ (mapM_ (\tyarg -> checkKindM (typeAnn tyarg) tyarg (Type ()))) namesAndTypes
```
And I'd probably use `for` here, as below.

effectfully replied:
> `mapM` is from before `Traversable` was in base, and has a needlessly strong `Monad` constraint

I agree, but then everybody uses `sequence`, even though by that logic we should use `sequenceA`.

michaelpj replied:
I agree that `sequenceA` is better :) I mostly find I just use `traverse`, though.

kwxm replied:
> Is... this using the traversable instance of `(,)`? I'd probably just not do that. How about:
> 
> ```
> traverse_ (\(_, tyarg) -> checkKindM ...) namesAndTypes
> ```

That was my first attempt (but with `mapM_` instead of `traverse`), but I thought I'd be told off for being too simplistic! It's quite a common pattern (which it took me a while to get used to) in the codebase to use `fmap` relying on the `Functor` instances of things on things like `(,)` and `Either`  (and indeed `Term`).  This seemed like a natural extension of that. I'lll go back to the explicit version though, since it's clearer what's going on.

michaelpj replied:
At least in my head, tuples are usually "data-like" things, and the fact that they have `Functor/Applicative/Monad` instances is kind of a quirk. I'm willing to fight for never using them (although no doubt I'm a hypocrite).

`Either` is a bit different, since it is actually widely used as a short-circuiting monad.

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -403,6 +455,24 @@ inferTypeM (Unwrap ann term)        = do
+inferTypeM (ApplyBuiltin ann bn typeArgs args) = do
+  TypeComponents typeVars argTypes rtype <- inferTypeComponentsOfBuiltinM ann bn
+  mapping <- checkBuiltinTypeArgs ann bn typeVars typeArgs              -- How to instantiate the type variables
+  normalizedArgTypes <- mapM (substNormalizeMappingM mapping) argTypes --  Instantiate and normalise the argument types
```
Do we check anywhere that each type in `typeArgs` (and also `rtype`) are of kind `*`? I guess it's guaranteed by the extensible types machinery, but I think we should perform a check here nonetheless, because the extensible types machinery may change in future and it makes sense to check kinds before normalizing something anyway.

kwxm replied:
`checkBuiltinTypeArgs` does the kind checking for the argument types, but we never check the kind of the return type, not even in the specification.  Should we do that?

I suppose that ideally the (formal) type parameters for a builtin should carry their kinds with them, even if they are all `*`.  I think we're aware that there's a slight mismatch here though.  There was a question of whether reworking the type schemes might help with this.

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -403,6 +455,24 @@ inferTypeM (Unwrap ann term)        = do
+
+-- [infer| bn : [a_1 :: K_1, ..., a_n :: K_m](B_1, ..., B_n)C]
+-- [check| G !- A_i :: K_i]
+-- [check| G !- M_j : T_j]     [A_i / a_i]* B_j ~>? S_j   T_j ~ S_j
```
I think you directly check `M_j` against `S_j` without introducing any intermediate `T_j`?

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -403,6 +455,24 @@ inferTypeM (Unwrap ann term)        = do
+-- operations from Normalize.Internal.
+
+-- [infer| bn : [a_1 :: K_1, ..., a_n :: K_m](B_1, ..., B_n)C]
+-- [check| G !- A_i :: K_i]
```
This is not what is happening right now, because currently we assume that all `K_i` are `*`, right?

kwxm replied:
Yes, but the specification allows higher kinds. I'm not sure what to do there.  This relates to the remark above. Maybe we should make the builtin signatures carry kinds as well so that we can ignore that issue here and localise the problem in one place.  At the moment we have a global assumption that everything has kind `*` and it would be easy to change that in one place and forget to fix it elsewhere.

effectfully commented on language-plutus-core/test/Evaluation/DynamicBuiltins/Definition.hs:
```diff
@@ -120,18 +110,21 @@ test_dynamicId =
+                                 [TyFun () integer integer]
+                                 [Apply () constIntegerInteger one]
+                            )
+                            zer
```
What happens if we write

```haskell
ApplyBuiltin ()
    (DynBuiltinName dynamicIdName)
    [TyFun () integer integer]
    [Apply () constIntegerInteger one, zer]
```

? Type error, I guess? "Too many arguments". Is this what we want? Probably yes, but pinging @michaelpj just in case.

michaelpj replied:
Too many arguments sounds right, and I expect to get that from the typechecker (which I think we will in this PR). Maybe worth some tests for the over/under application error cases.

kwxm replied:
I did make a mini-issue to add some tests at https://jira.iohk.io/browse/SCP-771 a while ago.  I was going to do that immediately after this is merged (if it ever is merged).

effectfully commented on language-plutus-core/src/Language/PlutusCore/View.hs:
```diff
@@ -5,65 +5,17 @@
 
 module Language.PlutusCore.View
```
No I think, because `IterApp` should be used in tests.

effectfully commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -403,6 +455,24 @@ inferTypeM (Unwrap ann term)        = do
+inferTypeM (ApplyBuiltin ann bn typeArgs args) = do
+  TypeComponents typeVars argTypes rtype <- inferTypeComponentsOfBuiltinM ann bn
+  mapping <- checkBuiltinTypeArgs ann bn typeVars typeArgs              -- How to instantiate the type variables
+  normalizedArgTypes <- mapM (substNormalizeMappingM mapping) argTypes --  Instantiate and normalise the argument types
```
`withExtendedTypeVarEnv` works in the `NormalizeTypeT` transformer and `checkBuiltinTermArgs` runs in the `TypeCheckM` monad, so you can't mix those two operations like that.

Maybe you can lift `checkBuiltinTermArgs` into  `NormalizeTypeT TypeCheckM` and then call `runNormalizeTypeM` (which should be called `runNormalizeTypeT`) outside of the whole computation, but that seems harder to comprehend. In general, we could do

```haskell
expandType <- substNormalizeMappingM <$> checkBuiltinTypeArgs ann bn typeVars typeArgs
normalizedArgTypes <- traverse expandType argTypes
checkBuiltinTermArgs ann bn normalizedArgTypes args
expandType rtype
```

at which point we can move `substNormalizeMappingM` into `checkBuiltinTypeArgs` and avoid creating any `TypeMapping` at all and just directly construct the function.

michaelpj replied:
Could we not also teach `runNormalizeTypeT` to take an initial environment? Or is that a weird thing to do?

reactormonk commented on language-plutus-core/budgeting-bench/csvs/benching.csv:
```diff
@@ -9371,3 +9371,300 @@ VerifySignature/ExMemory 1024,1.9304307891148622e-5,1.9299324972813317e-5,1.9309
 VerifySignature/ExMemory 2048,1.9319871745801975e-5,1.9303470563033934e-5,1.937191089766017e-5,9.467913018862859e-8,2.6851240364915522e-8,1.9448109481905572e-7
 VerifySignature/ExMemory 4096,1.92667551061035e-5,1.925945208676366e-5,1.9272958060574025e-5,2.171057215677812e-8,1.72039262683514e-8,2.6643010985045755e-8
 VerifySignature/ExMemory 8192,1.9249784997943423e-5,1.9241689703204302e-5,1.925933683096966e-5,3.1382349142082264e-8,2.520129209534575e-8,4.12958304765296e-8
+Name,Mean,MeanLB,MeanUB,Stddev,StddevLB,StddevUB
```
Yeah, some rerun the bench binary. It doesn't delete the original file when you run it a second time, criterion behavior. Maybe I should change that?

kwxm replied:
That must have been me.  I'll revert it.

kwxm replied:
If and when we merge it, @reactormonk should make sure it's doing costing/benchmarking sensibly.

reactormonk commented on language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExMemory.hs:
```diff
@@ -102,15 +101,20 @@ instance (Generic a, GExMemoryUsage (Rep a)) => ExMemoryUsage (GenericExMemoryUs
+-- FAO @reactorkonk: This is needed to deal with the lists in ApplyBuiltin;
+-- There was an existing instance for String which had to be deleted.
+instance ExMemoryUsage a => ExMemoryUsage [a] where
+    memoryUsage l = sum $ fmap (\e -> memoryUsage e + 1) l
```
The `+ 1` should be left explicitly here, we might want to adjust that later. It's also not "next number", but "additional cost", so the `succ` IMO doesn't fit the semantics. Agree on the `Monoid` part.

effectfully replied:
> The + 1 should be left explicitly here, we might want to adjust that later. It's also not "next number", but "additional cost", so the succ IMO doesn't fit the semantics.

Sure, we can have `mappend additionalCost` or whatever else in place of `succ`.

kwxm commented on language-plutus-core/src/Language/PlutusCore/Constant/Apply.hs:
```diff
@@ -62,32 +53,35 @@ applyTypeSchemed
                     (ExcessArgumentsConstAppError $ void <$> args)
-                    Nothing
+                    (Just $ ApplyBuiltin () name [] [])
+                  --  ^ This should really contain the type and term arguments, but we don't have them here.
```
That would mean adding an extra argument to quite a lot of functions: this code is in `applyTypeSchemed`, which is called in `applyTypedBuiltinName` (and elsewhere), which is called by `applyBuiltinName` (and elsewhere), and so on.  We could add an extra argument for the sake of a precise error message, but is it worth it? I wonder if it would increase exeuction costs?

kwxm commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -302,18 +316,62 @@ unfoldFixOf pat arg k = do
+    case zipExact tys args of
+      Left LeftLonger  -> throwError $ BuiltinTypeErrorE ann (BuiltinUnderApplied bn)
+      Left RightLonger -> throwError $ BuiltinTypeErrorE ann (BuiltinOverApplied  bn)
+      Right argsAndTypes -> mapM_ checkArg argsAndTypes
```
Right: that's a bit neater. I've added `for_` and `traverse` to `PlutusPrelude`.

kwxm commented on language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs:
```diff
@@ -302,18 +316,62 @@ unfoldFixOf pat arg k = do
+      Right namesAndTypes ->  do
+        mapM_ (mapM_ (\tyarg -> checkKindM (typeAnn tyarg) tyarg (Type ()))) namesAndTypes
+        -- NB: All of the type variables in TypeSchemes have kind *  ^
+        mapping <- mapM (mapM (normalizeType . void)) namesAndTypes  -- ie, normalise all the types in the (TyName, Type) pairs
```
Hmm. I'm not so sure about that.  For the final line, I _think_ the best I can do is
```
 for namesAndTypes $ \(tyName,tyArg) -> do
                     tyArg' <- normalizeType . void $ tyArg
                     pure (tyName, tyArg')
```
Maybe just having `mapM` or `traverse` with an explanation would be simpler?  Doing the normalisation before the zipping somehow seems a bit weird as well.  We then do a kind check on the normalised type, and it feels like tht sould be done prior to normalisation.  However, I'll experiment with using extended environments rather than these `TypeMapping` things and see what direction that pushes me in.

michaelpj replied:
I think what you wrote is better. I also don't think it's weird to do the normalization before zipping.

kwxm commented on language-plutus-core/prelude/PlutusPrelude.hs:
```diff
@@ -182,3 +184,15 @@ printPretty = print . pretty
+zipExact :: [a] -> [b] -> Either LengthMismatch [(a,b)]
+zipExact = zipExact' []
+     where zipExact' acc [] []         = Right $ reverse acc
+           zipExact' acc (a:as) (b:bs) = zipExact' ((a,b):acc) as bs
```
Pure ignorance!  That didn't even cross my mind. I'm old-fashioned and in the habit of using accumulators from OCaml programming.  Mind you, I'm never quite sure how accumulators behave when laziness is involved.

kwxm commented:
Re Eq.hs, @michaelpj says (but for some reason I can't reply directly to the comment)

> Are we going to improve this now, or later?

Can we do it later, in a separate PR?  I think that's kind of orthogonal to the main business here.

kwxm commented:
I think we can safely close this now.
